src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -277,7 +285,7 @@ namespace ts.codefix {
         if (isCallExpression(node) && hasPropertyAccessExpressionWithName(node, "then") && nodeType && !!transformer.checker.getPromisedTypeOfPromise(nodeType)) {
             return transformThen(node, transformer, outermostParent, prevArgName);
         }
-        else if (isCallExpression(node) && hasPropertyAccessExpressionWithName(node, "catch") && nodeType && !!transformer.checker.getPromisedTypeOfPromise(nodeType)) {
+        else if (isCallExpression(node) && hasPropertyAccessExpressionWithName(node, "catch") && nodeType && !!transformer.checker.getPromisedTypeOfPromise(nodeType) && (!prevArgName || "identifier" in prevArgName)) {

```

Comment:
Is there a particular reason we fail if we see a catch but the `prevArgName` is a binding element?

Reply:
It doesn't actually fail, it falls through to
```ts
else if (nodeType && transformer.checker.getPromisedTypeOfPromise(nodeType)) {
  return transformPromiseCall(node, transformer, prevArgName);
}
```
and basically becomes a no-op. This happens only in some really specific structures where a `then` and a `catch` both return a value and then another `then` destructures that value. The result is that the outer promise gets transformed and the inner, nested one doesn't. (And really, the resulting code is more concise than the fully transformed async version in the examples I've been able to track down.) But strictly speaking, this is an incomplete solution. Let me give it another hour or two and see if I can come up with something better.

Reply:
Got it! e7fb18e395b3e0874c22d96b0a2c94fec2eac1d0 now handles this case, and you can see the corresponding test case [here](https://github.com/Microsoft/TypeScript/pull/31008/files#diff-d2f0b9a35c740584b0f344544068dd72R23)

Reply:
Looks like we're missing the baseline file for it?

Reply:
Hm, so we are. Updated.